tensorflow/python/distribute/cross_device_utils.py

```diff
@@ -557,46 +557,38 @@ def all_reduce_indexed_slices(self,
     # have consistent length across the board, we handle the reduction of
     # IndexedSlices as follows:
     #   1. Gather the lengths of IndexedSlices from all participants.
-    #   2. If they have consistent length, apply all_gather.
-    #   3. Otherwise convert IndexedSlices to dense tensors and apply
-    #      all_reduce.
+    #   2. If they do not have consistent length, they are padded with zeros to the same length.
+    #   3. Apply all gather.
     with ops.device(self._device):
-
-      def all_gather():
-        """Use all_gather to aggregate `IndexedSlices`."""
-        all_values = self._all_gather(
-            input_slices.values, communication_hint, timeout=timeout)
-        # Add control dependency to order the all-gather.
-        control = [all_values] if communication_hint == 'NCCL' else []
-        with ops.control_dependencies(control):
-          all_indices = self._all_gather(
-              input_slices.indices, communication_hint, timeout=timeout)
-        return ops.IndexedSlices(
-            values=all_values,
-            indices=all_indices,
-            dense_shape=input_slices.dense_shape)
-
-      def densify_and_all_reduce():
-        """Use all_reduce to aggregate `IndexedSlices`."""
-        densified = ops.convert_to_tensor(input_slices)
-        reduced = self.all_reduce(
-            densified, communication_hint=communication_hint, timeout=timeout)
-        # We have to convert dense grad to IndexedSlice because all_reduce()
-        # and all_gather() must have the same return type as required by
-        # control_flow_ops.cond.
-        return ops.IndexedSlices(
-            values=reduced,
-            indices=math_ops.range(array_ops.shape(reduced)[0]),
-            dense_shape=input_slices.dense_shape)
-
       length = array_ops.shape(input_slices.indices)
       all_lengths = self._all_gather(
```

Comment:
Can we use self.gather which already contains the padding logic? I understand that self.gather() is not most efficient in theory, but we can always optimize it if necessary and it can benefit more use cases